### PR TITLE
Issue5164 [WIP]

### PR DESCRIPTION
**Pull request by P4N74**
Ability to seek+read is yet to be fixed.

### CODE REVIEW

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -2371,7 +2371,10 @@ struct Elf_(r_bin_elf_obj_t)* Elf_(r_bin_elf_new_buf)(RBuffer *buf) {
 	struct Elf_(r_bin_elf_obj_t) *bin = R_NEW0 (struct Elf_(r_bin_elf_obj_t));
 	bin->kv = sdb_new0 ();
 	bin->b = r_buf_new ();
-	bin->size = buf->length;
+//	bin->size = buf->length;
+	ut64 curr = lseek (buf->fd, 0, SEEK_CUR);
```
Comment (radare) [inline libr/bin/format/elf/elf.c]: it is a bad idea to use buf->fd, or even lseek directly. those vars are suposed to be private to the abstraction of RBuffer.  an rbuffer can even work without an fd, so it may segfault or produce wrong results.

File: libr/bin/p/bin_elf.c [hunk 2]
```diff
@@ -47,6 +47,7 @@ static void * load_bytes(RBinFile *arch, const ut8 *buf, ut64 sz, ut64 loadaddr,
 		return NULL;
 	}
 	tbuf = r_buf_new ();
+	tbuf->fd = arch->fd;
```
Comment (radare) [inline libr/bin/p/bin_elf.c]: better not to force this here. we should work with rbuffers independely from the data source baked behind the implementation

File: libr/io/io.c [hunk 3]
```diff
@@ -1007,16 +1007,21 @@ static ut8 *r_io_desc_read(RIO *io, RIODesc *desc, ut64 *out_sz) {
...
-		return NULL;
+		eprintf ("WARNING: File is greater than 0x%"PFMT64x" bytes.\nAllocating R_IO_MAX_ALLOC set as the environment variable.\n", io->maxalloc);
+		*out_sz = io->maxalloc;
 	}
 
 	buf = malloc (*out_sz);
 	if (!buf) {
-		eprintf ("Cannot allocate %" PFMT64d " bytes\n", *out_sz);
-		return NULL;
+		if (*out_sz > R_IO_MAX_ALLOC) {
```
Comment (radare) [inline libr/io/io.c]: maybe >=

File: libr/io/io.c [hunk 4]
```diff
@@ -1007,16 +1007,21 @@ static ut8 *r_io_desc_read(RIO *io, RIODesc *desc, ut64 *out_sz) {
...
+			eprintf ("Failed to allocate %" PFMT64u " bytes.\nAllocating %" PFMT64u " bytes.\n", *out_sz, R_IO_MAX_ALLOC);
```
Comment (radare) [inline libr/io/io.c]: as long as allocating can result in really huge numbers, maybe we should use r_num_units() to get a string represented in KB, MB, GB, ...

File: libr/bin/format/elf/elf.c (hunk 1 above)
Comment (P4N74) [inline libr/bin/format/elf/elf.c]: Agreed that RBuffer can work without an fd, but in this case, it is called from http://computers.tutsplus.com/tutorials/how-to-ocr-documents-for-free-in-google-drive--cms-20460 which sets this. But as you commented there that we should have independent working of rbuff, in that case, it seems like the current way I am trying to implement won't be doable without the fd since I am trying to do seek+read in the r_buf_fcpy_at part to which only the RBuffer is passed and no other related info about the bin (esp. file size and fd to read from).

Comment (radare) [inline libr/bin/format/elf/elf.c]: can you squash the commit with just the changes of int->ut64 to see if that breaks something? i would put all that rbuf mangling in a separate pr, to ease the review of broken tests.

this should be solved by adding an api in RBuffer to get the size instead of directly accesing the ->length field. rbuffer can do the lseek in case fd != -1 or return length otherwise, etc..

File: libr/io/io.c (hunk 4 above)
Comment (P4N74) [inline libr/io/io.c]: this part is already present in the code. here: https://github.com/radare/radare2/blob/master/libr/io/io.c#L43

**Comment by radare** [comment]
theres one test failing

File: libr/io/io.c (hunk 4 above)
Comment (radare) [inline libr/io/io.c]: I dont see maxalloc in thw code block you are refering to

**Comment by radare** [comment]
screen shot 2016-07-02 at 11 51 12

File: libr/io/io.c (hunk 4 above)
Comment (P4N74) [inline libr/io/io.c]: `io->maxalloc = r_num_get (NULL, env);` this part get's the env variable and represents it in terms of bytes. What I understand from you comment is that user can write `R_IO_MAX_ALLOC=128M` and it should be converted into proper bytes. This is done at this place.

Comment (radare) [inline libr/io/io.c]: No i was not refering to the numget. But rather to the eprintf showing the sizes in decimal instead of unit format

Comment (P4N74) [inline libr/io/io.c]: oh. Sure. I will update it. Currently fixing the failing test case.

Comment (radare) [inline libr/io/io.c]: Ok good!

**Comment by radare** [comment]
ping?

**Comment by P4N74** [comment]
Hi, it's  addition of `(section->size + section->paddr) > sz1` in place of `size == 0 || size > 0xf0000000` condition which is giving the error. Haven't been able to figure out the exact reason.

File: libr/io/io.c (hunk 4 above)
Comment (P4N74) [inline libr/io/io.c]: Done

File: libr/core/bin.c [hunk 5]
```diff
@@ -1548,7 +1548,7 @@ static int bin_sections(RCore *r, int mode, ut64 laddr, int va, ut64 at, const c
 			}
 			r_meta_add (r->anal, R_META_TYPE_COMMENT, addr, addr, str);
 			if (section->add)
-				r_io_section_add (r->io, section->paddr, addr, section->size, section->vsize, section->srwx, section->name, 0, fd);
+					r_io_section_add (r->io, section->paddr, addr, section->size, section->vsize, section->srwx, section->name, 0, fd);
```
Comment (alvarofe) [inline libr/core/bin.c]: wrong ident

File: libr/io/io.c [hunk 6]
```diff
@@ -1007,16 +1007,27 @@ static ut8 *r_io_desc_read(RIO *io, RIODesc *desc, ut64 *out_sz) {
...
 	}
 
 	buf = malloc (*out_sz);
 	if (!buf) {
-		eprintf ("Cannot allocate %" PFMT64d " bytes\n", *out_sz);
-		return NULL;
+		if (*out_sz > R_IO_MAX_ALLOC) {
+			char *num_unit = r_num_units (NULL, *out_sz);
+			eprintf ("Failed to allocate %s bytes.\n"
+					"Allocating %" PFMT64u " bytes.\n", num_unit, R_IO_MAX_ALLOC);
```
Comment (alvarofe) [inline libr/io/io.c]: cast R_IO_MAX_ALLOC to ut64 since #define by default use int

File: libr/io/section.c [hunk 7]
```diff
@@ -57,8 +57,8 @@ static RIOSection *findMatching (RIO *io, ut64 paddr, ut64 vaddr, ut64 size, ut6
 R_API RIOSection *r_io_section_add(RIO *io, ut64 offset, ut64 vaddr, ut64 size, ut64 vsize, int rwx, const char *name, ut32 bin_id, int fd) {
 	int update = 0;
 	RIOSection *s;
-	if (size == 0 || size > 0xf0000000) {
-		if (size > 0 && size != UT64_MAX && size != UT32_MAX)
+	if (size == 0) {
```
Comment (alvarofe) [inline libr/io/section.c]: to me is more elegant `if (!size)` instead of `if (size == 0)`

File: libr/io/section.c [hunk 8]
```diff
@@ -57,8 +57,8 @@ static RIOSection *findMatching (RIO *io, ut64 paddr, ut64 vaddr, ut64 size, ut6
...
+		if (size > 0 && size != UT64_MAX && size != UT32_MAX) //hacky things which might give bad output in case size == UT32_MAX for 64bit elf.
```
Comment (alvarofe) [inline libr/io/section.c]:
```C
	if (size == 0) {
		if (size > 0 && size != UT64_MAX && size != UT32_MAX) //hacky things which might give bad output in case size == UT32_MAX for 64bit elf.
			eprintf ("Invalid size (0x%08" PFMT64x
				 ") for section '%s' at 0x%08" PFMT64x "\n",
				 size, name, vaddr);
		return NULL;
	}

```

you only will reach that statement when `size == 0` and then `size > 0` will never be true unless i am missing something. I think it should be

```C
	if (!size && size == UT64_MAX && size == UT32_MAX) //hacky things which might give bad output in case size == UT32_MAX for 64bit elf.
		eprintf ("Invalid size (0x%08" PFMT64x
			 ") for section '%s' at 0x%08" PFMT64x "\n",
			 size, name, vaddr);
	return NULL;
```

**Comment by radare** [comment]
Please update the PR and i'll merge. it seems to make travis happy and doesnt seems to break anything. Cool!

File: libr/io/section.c [hunk 9]
```diff
@@ -57,11 +57,12 @@ static RIOSection *findMatching (RIO *io, ut64 paddr, ut64 vaddr, ut64 size, ut6
 R_API RIOSection *r_io_section_add(RIO *io, ut64 offset, ut64 vaddr, ut64 size, ut64 vsize, int rwx, const char *name, ut32 bin_id, int fd) {
 	int update = 0;
 	RIOSection *s;
-	if (size == 0 || size > 0xf0000000) {
-		if (size > 0 && size != UT64_MAX && size != UT32_MAX)
+	if (!size || size == UT64_MAX || size == UT32_MAX) { //hacky things which might give bad output in case size == UT32_MAX for 64bit elf. Check on basis of size, offset and file size would be a good idea.
```
Comment (P4N74) [inline libr/io/section.c]: @alvarofe for the comment: https://github.com/radare/radare2/pull/5242#discussion_r69490005 , I have updated the condition. The eprintf was supposed to run when size is greater than 0xf0000000 but less than UT64_MAX and UT32_MAX. Now since `size > 0xf0000000` is no longer checked, I just removed the eprintf part and updated the main condition to fail in case of either of above condition.

**Comment by radare** [comment]
merged! looks good to me now :)
